graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -51,14 +51,26 @@ std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> UniqueAndCompact(
         // If the given num_bits argument is not in the reasonable range,
         // we recompute it to speedup the expensive sort operations.
         if (num_bits <= 0 || num_bits > sizeof(scalar_t) * 8) {
-          auto max_id = thrust::reduce(
-              exec_policy, src_ids_ptr, src_ids_ptr + src_ids.size(0),
-              static_cast<scalar_t>(0), thrust::maximum<scalar_t>{});
-          max_id = thrust::reduce(
+          auto max_id_src = torch::empty(
```

Comment:
Are you re-implementing the max operation?
- why cuda does not provide the optimal op?
- what about move it out as a util function?

Reply:
`thrust::reduce` synchronizes the GPU with the CPU. I am refactoring the pattern I use here into CopyScalar so that I don't have to allocate pinned tensors myself anymore. That is why, I re-marked this PR as a draft.

Reply:
I will work on it a little more and ping you when it is ready to review again.

Reply:
Refactored the max operator and refactored and optimized the whole implementation. Could you take another look?